### PR TITLE
Estimator Head design document

### PR DESCRIPTION
**Pull request by ewilderj**
RFC document for design review pilot: Estimator Head API design.

(*Not to be merged into tensorflow/ - will be merged into new community repo.)

Please leave design feedback in this PR. Comment period will be open until 2018-05-22.

cc @ispirmustafa @roumposg

## Summary

| Status        | Proposed                                          |
:-------------- | :-------------------------------------------------|
| **Author(s)** | George Roumpos (Google), TensorFlow team          |
| **Sponsor**   | Mustafa Ispir (Google)                            |
| **Updated**   | 2018-05-01                                        |

In this doc we discuss the Head API, which helps users define customizable
models that follow the `tf.estimator` API. The API
includes:

*   A `Head` interface.
*   Factory methods to create common heads, such as regression head.
*   A `multi_head` method that combines more than one heads for multi-objective
    learning.
*   Canned estimators that can use those heads.

The API is already exposed under `tf.contrib.estimator`. The code is in
`tensorflow/contrib/estimator/python/estimator/head.py`,
`tensorflow/contrib/estimator/python/estimator/multi_head.py` and
`tensorflow/python/estimator/canned/head.py`, and an earlier (deprecated)
version is exposed under `tf.contrib.learn`. The goal of this design doc is to
graduate the API from contrib to core.

### CODE REVIEW

**Comment by eaplatanios** [comment]
Thank you very much for creating this process and being open to community feedback! :)

I like the head API proposal but I would like to add something to it, based on my experience of trying to create a high-level learning API on TensorFlow Scala.

We usually want different inputs while training and while performing inference (e.g., target labels in supervised learning are only provided while training). However, we also might want to produce different outputs while training and while performing inference. For example, in a machine translation task, I might want to provide a single distribution over tokens for each time point while training, but a batch of complete sequence predictions along with their scores, while performing inference. I have implemented an interface that supports those use cases too in my API, but I'm not very happy with it and so I would be interested in adopting the head API too. However, I think that this use case is not currently supported here. Is that true? If so, would you be open to a discussion about adding support for this use case?

Note that these different heads that I am describing, also require different inputs to be fed to the model. In the current design proposal all heads are provided the same inputs, right?

Out of curiosity, was this API inspired from the Tensor2Tensor library?

Thanks,
Anthony

**Comment by roumposg** [comment]
Hi @eaplatanios,
As @ispirmustafa said, the API supports different inputs for training and infrerence, as well as different outputs. E.g. during training the output is typically the loss, whereas for inference we can return e.g. probabilities for all classes, or top-k classes.

But you mention the case of sequence models, such as translation. We did not touch this use case yet. Do you think the API is compatible with those models? If not, we are interested about what are the specific blockers.

**Comment by ngc92** [comment]
The Head API looks very nice for reducing boilerplate when writing custom model functions. 
My two observations:
1) There does not currently seem to be a way to use different loss weights for different labels (e.g. if you have a very imbalanced dataset. I am not sure how well that would fit into the `LossSpec` interface.
2) Suppose I wanted to implement a general `AdversarialTrainingHead` that takes an existing Head as input but modifies the loss function to also include the loss for an adversarial modification of the input, so its loss function might look something like 
```
loss = original_head.create_loss(logits, labels)
adv = make_adversarial_perturbation()
adv_loss = original_head.create_loss(logit_fn(adv), labels)
return loss + adv_loss
```
This would be very difficult to implement in the current setting because even though the design doc mentions a `logit_fn` the `create_estimator_spec` actually takes a pre-existing Tensor. 
For many use cases this seems like the more convenient choice, but maybe it would make sense to allow passing in an actual `logit_fn` together with `features`.

**Comment by DavidNorman** [comment]
Will there be a way for explicitly targeting a particular device?

For instance:  if in the `logits_fn` I explicitly set the device target (`tf.device()`), then will any automatically generated loss and training networks follow that target?

**Comment by roumposg** [comment]
Hi @ngc92,
1. The head factory methods we provide offer a weight_column argument. You can then pass the weights in the features dict, and have a different weight per example. Would this cover your use case?
2. For adversarial training, we recommend multi_head. One head will have the loss related to your objective function, the second head will have the adversarial loss. Multi_head will sum the two losses. Is this what you are looking for?

Hi @DavidNorman,
I think the following will work:

```python
with tf.device(...):
  logits = my_logit_fn()
  estimator_spec = my_head.create_estimator_spec()
```

**Comment by ngc92** [comment]
@roumposg 
1. It will work if there is only one class per example (i.e. multi_class_head) but in multi_label_head one example could belong to multiple classes, so these two weighting concepts become orthogonal. If the weight_column could have shape `BATCH_SIZE, NUM_CLASSES` one could do this more fine-grained weighting, but this would make `LossSpec`'s `unreduced_loss` and `weights` also be of shape `BATCH_SIZE, NUM_CLASSES`. 
Of course the user could also provide a custom loss function that does the weighting as desired, so this point is not about what is possible with the API but about what the predefined Heads support.

**Comment by roumposg** [comment]
@ngc92 
Yes, if you have 2D weight matrix, the recommended way is to define your custom `loss_fn`. The reason we did that is to avoid surprises related to the `loss_reduction` argument. E.g. some users may expect to sum over the batch but average over the classes, and it was not clear to us how to support all use cases in a clean way.

Thanks for the feedback. I see that this is not well explained, we will explain in better in a future version of the document.

**Comment by bhack** [comment]
@ewilderj Can we add a special label for design review/RFC issues/PRs like this one?

**Comment by bhack** [comment]
How this will interact with the `tf.keras.estimator.model_to_estimator` use case?

**Comment by ngc92** [comment]
@roumposg
Ok, so I've spent some time thinking about how to implement an adversarial head with the current design (so that you could then do a MultiHead to combine it with your regular training).

1. A minimal way of doing adversarial training would be to just add the adversarial loss to your `Head` as an additional regularization loss. However you probably also would want additional eval statistics and eval metrics (i.e. the accuracy on adversarial data) which you would have to add by hand.

2. Using the same head twice. An adversarial training `model_fn` with the current design could look like this:
```
logits = logit_fn(features)
adv_f = make_adversarial(logits, features)
adv_l = logit_fn(adv_f)
base_head = MyHead()
adv_head = MyHead(name="adversarial")
head = MultiHead(base_head, adv_head)
return head.make_estimator_spec(..., logits={"head": logits, "adversarial": adv_l})
```

3. The documentation mentions a split of a model_fn into a head and a logit_fn, which does not seem to be really reflected in the current design. If `make_estimator_spec` took an actual `logit_fn` instead of a logit tensor, one could write a head like
```
class FGSM_AdvHead(Head):
  def __init__(self, base_head):
     ....
  def make_estimator_spec(features, logit_fn, ...):
    logits = self.logit_fn(features)
    adv_f = self.make_adversarial(logits, features)
    adv_l = self.logit_fn(adv_f)
    head = MultiHead(self.base_head(name="base"), self.base_head(name="adv"))
    return head.make_estimator_spec(..., logits={"base": logits, "adv": adv_l})
```
where a real implementation would probably do something smarter/more specialized than constructing a temporary MultiHead.

This last option is currently no possible, right?

So how could this be fixed?
1) Simply state that this is out of scope. Note that this disables to write any head that wants to control the context in which the `logit_fn` is called, or wants to call `logit_fn` multiple times. The most direct example I could come up with is the AdversarialHead above, but one might also be interested to call `logit_fn` for parts of the current batch on different devices, and maybe seq2seq models might also benifit from receiving their input in a functional form instead of a tensors (I haven't really thought about this last point, though). In this case I would rewrite the following paragraph in the documentation:

> Head offers an intermediate level of abstraction: the model_fn is split into a logit_fn that contains the ops to create the logits Tensor, and a Head that produces ops for train, eval and infer given those logits. Internally, canned estimators use this scheme. E.g. DNNClassifier and DNNRegressor use the same logit_fn but the former uses a classification head, whereas the latter uses a regression head. This reduces code duplication, improves testing, and makes it easier for users to adapt models for their own use.
 
at least to me this would insinuate that a usage as above described be possible, instead of this being just the implementation strategy of DNNClassifier/DNNRegressor

2) Take a `logit_fn` instead of logits. This would cause three problems:
 * The user might be required to write another function even though for most use cases, if the non-head part of the model_fn does something apart from calculating the logits. This could be fixed by making the interface take either a logit tensor or a function, where fany heads might only work with a function and otherwise raise an exception. 
 * In the current setting `make_estimator_spec` also takes regularization losses. These will depend on the calculation of the logits. Therefore one would need the `logit_fn` to be able to return a `(logits, regularization_loss)` tuple or more generally a `LogitSpec` object.
* In MultiHead in most cases one would not like each sub-head to calculate the same logits again. However, MultiHead might amend the `logit_fn` with some form of LRU cache that only adds ops to the graph if it is called with different features. (Maybe the graph optimizer would be able to fold all logit_fn subgraphs back together?)

If you do pass in the logits as a function instead of a tensor, this might also allow to lift the Head completely out of `model_fn` and provide a `make_estimator` instead of a `make_estimator_spec` function.

**Comment by roumposg** [comment]
@ngc92 thanks for the feedback.
I think option 2 for adversarial training works. You can also use the `head_weights` argument of `multi_head` to multiply the loss for each head. E.g. for the adversarial head you typically want to choose a negative head weight.

What we mean by splitting into head and logit_fn is that you can write your method that calculates logits, and pass those logits to the head. I don't think that accepting a logit_fn argument would make the api more flexible. Is there a use case that is not supported in your opinion?

**Comment by ngc92** [comment]
@roumposg
well, since you can easily combine higher and lower level interfaces in the `model_fn` I don't think this is something that is not supported by the current Head design.

I am not quite sure whether a `logit_fn` should be supported as it adds quite a bit of complexity. If Head only takes logit tensors, though, I would not call the calculation of that a `logit_fn`, as this (to me) sounds very much like the ability of passing a `model_fn` to an estimator.

I think to a user just writing a model option 2 from my last comment is fine, but from a library perspective (e.g. in case of the adversarial training for [cleverhans](https://github.com/tensorflow/cleverhans)) the third option seems more appealing. In that way passing a `logit_fn` instead of logits does increase what you can do as a writer of a general purpose head.

**Comment by ewilderj** [comment]
After design review, this RFC has been merged into the new community repo. See https://github.com/tensorflow/community/blob/master/rfcs/20180501-estimator-head.md

Thank you to everyone who participated in the discussion!

**Comment by yupbank** [comment]
@roumposg  would it make more sense to add a `label_sep` to `multi_label_head` so that `labels` don't need to be `sparse_tensor`  but a tensor?
  as here https://github.com/tensorflow/tensorflow/pull/20001/files#diff-eb023bd59df93d51913b23161f6187baR1457

**Comment by roumposg** [comment]
@yupbank `multi_label_head` is for problems where the number of labels is not fixed. So it makes sense to have `SparseTensor`. Do you have a problem with a fixed number of labels `n>1`? Then one option is to split each each example into `n` examples and use `multi_class_label`.

**Comment by yupbank** [comment]
@roumposg yeah, i understand that, but if we can have a labels feature column containing dimension 1 value, with `["label1=label2", "label2=label5"] ` and user can specify `=` as the `label_separator`,  the `multi_label_head` interface would be more consistent with other heads, which accept dense tensor

**Comment by roumposg** [comment]
Good idea. But this is easy to implement using `tf.string_split` in your `input_fn`. I think adding one more argument to the api will add complexity without adding much value. What do you think? If you want to use a dense `Tensor`, `multi_label_head` supports multi-hot tensor of shape `[batch_size, n_classes]`.

**Comment by yupbank** [comment]
i mean it would add value for users that, they can easily switch to one vs all mode, using `multi_class_head` without need to change the label tensor.

**Comment by ispirmustafa** [comment]
Peng, you can set loss_fn argument of `multi_class_head`. We can add an
example to the documentation of `multi_class_head`. What do you think?
